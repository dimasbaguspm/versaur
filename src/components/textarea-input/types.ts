import type { TextareaHTMLAttributes, ReactNode } from 'react'

/**
 * Props for the TextAreaInput component
 */
export interface TextAreaInputProps
  extends Omit<TextareaHTMLAttributes<HTMLTextAreaElement>, 'size'> {
  /**
   * Visual style variant supporting Versaur color system
   * Core variants: primary (coral), secondary (sage), tertiary (mist), ghost (slate), neutral (light gray)
   * Semantic variants: success, info, warning, danger
   * Each variant supports outline form for flexible design expression
   */
  variant?:
    | 'primary'
    | 'primary-outline'
    | 'secondary'
    | 'secondary-outline'
    | 'tertiary'
    | 'tertiary-outline'
    | 'ghost'
    | 'ghost-outline'
    | 'neutral'
    | 'neutral-outline'
    | 'success'
    | 'success-outline'
    | 'info'
    | 'info-outline'
    | 'warning'
    | 'warning-outline'
    | 'danger'
    | 'danger-outline'
  /**
   * Label text to display above the textarea
   */
  label: ReactNode
  /**
   * Helper text to display below the textarea
   */
  helperText?: ReactNode
  /**
   * Error message for invalid state
   */
  error?: ReactNode
  /**
   * Minimum number of rows for the textarea
   */
  minRows?: number
  /**
   * Maximum number of rows for the textarea
   * Only applies when fieldSizing is 'fixed'
   */
  maxRows?: number
  /**
   * Enable field-sizing CSS property for auto-resizing
   * 'content' - Auto-resize to fit content (ignores maxRows)
   * 'fixed' - Fixed size with manual resize handle (respects minRows/maxRows)
   */
  fieldSizing?: 'content' | 'fixed'
}
