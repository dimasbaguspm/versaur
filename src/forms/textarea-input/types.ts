import type { HTMLAttributes, ReactNode } from 'react'

/**
 * Props for the TextAreaInput component
 */
export interface TextAreaInputProps
  extends Omit<
    HTMLAttributes<HTMLDivElement>,
    'onChange' | 'children' | 'defaultValue'
  > {
  /**
   * Label text to display above the textarea
   */
  label?: ReactNode
  /**
   * Helper text to display below the textarea
   */
  helperText?: ReactNode
  /**
   * Error message for invalid state
   */
  error?: ReactNode
  /**
   * The current value of the textarea (for controlled components)
   */
  value?: string
  /**
   * The default value for uncontrolled components
   */
  defaultValue?: string
  /**
   * Callback fired when the content changes
   */
  onChange?: (value: string) => void
  /**
   * Whether the textarea is disabled
   */
  disabled?: boolean
  /**
   * Whether the textarea is read-only
   */
  readOnly?: boolean
  /**
   * Placeholder text to display when empty
   */
  placeholder?: string
  /**
   * Name attribute for form submission
   */
  name?: string
  /**
   * Number of rows (height in rem units)
   * @default 3
   */
  row?: number
  /**
   * Whether the textarea is required
   */
  required?: boolean
}
