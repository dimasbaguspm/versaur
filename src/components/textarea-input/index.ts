export { TextAreaInput } from './textarea-input'
export type { TextAreaInputProps } from './types'
