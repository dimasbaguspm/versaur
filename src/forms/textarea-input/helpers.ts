import { cva } from '@/utils/variants'
import type React from 'react'

export const textAreaInputVariants = cva(
  'block w-full rounded-md border bg-white transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:pointer-events-none disabled:bg-gray-50 overflow-y-auto whitespace-pre-wrap break-words',
  {
    variants: {
      state: {
        default:
          'border-primary/30 text-foreground focus:border-primary focus:ring-primary/20',
        error:
          'border-danger bg-danger/5 text-foreground focus:border-danger focus:ring-danger/20',
        readOnly:
          'border-gray-300 bg-gray-50 text-foreground cursor-default focus:ring-0 focus:border-gray-300',
      },
    },
    defaultVariants: {
      state: 'default',
    },
  }
)

/**
 * Get the appropriate state for the textarea based on props
 */
export const getTextAreaState = (
  readOnly?: boolean,
  hasError?: boolean
): 'default' | 'error' | 'readOnly' => {
  if (readOnly) return 'readOnly'
  if (hasError) return 'error'
  return 'default'
}

/**
 * Handle paste event to insert plain text only
 */
export const handlePlainTextPaste = (
  e: React.ClipboardEvent<HTMLDivElement>,
  disabled?: boolean,
  readOnly?: boolean
) => {
  // Prevent paste when disabled or readOnly
  if (disabled || readOnly) {
    e.preventDefault()
    return
  }

  // Paste as plain text using modern API
  e.preventDefault()
  const text = e.clipboardData.getData('text/plain')

  // Use the modern Selection API instead of deprecated execCommand
  const selection = window.getSelection()
  if (!selection || !selection.rangeCount) return

  // Delete the current selection if any
  selection.deleteFromDocument()

  // Insert the plain text at the current position
  const range = selection.getRangeAt(0)
  const textNode = document.createTextNode(text)
  range.insertNode(textNode)

  // Move cursor to end of inserted text
  range.setStartAfter(textNode)
  range.setEndAfter(textNode)
  selection.removeAllRanges()
  selection.addRange(range)

  // Trigger input event to update state
  const inputEvent = new Event('input', { bubbles: true })
  e.currentTarget.dispatchEvent(inputEvent)
}
