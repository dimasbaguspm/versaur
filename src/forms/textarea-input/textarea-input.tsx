import React from 'react'
import { cn } from '@/utils/cn'
import {
  textAreaInputVariants,
  getTextAreaState,
  handlePlainTextPaste,
} from './helpers'
import type { TextAreaInputProps } from './types'

/**
 * TextAreaInput component for Versaur UI
 *
 * Uses contentEditable div for robust multi-line text input with better control
 * Provides error states, disabled, and readOnly support
 * Follows browser standards and accessibility best practices
 */
export const TextAreaInput = React.forwardRef<
  HTMLDivElement,
  TextAreaInputProps
>(
  (
    {
      label,
      helperText,
      error,
      className,
      disabled,
      readOnly,
      id,
      value,
      defaultValue,
      onChange,
      placeholder,
      name,
      row = 3,
      required,
      onBlur,
      onFocus,
      onKeyDown,
      ...props
    },
    ref
  ) => {
    const generatedId = React.useId()
    const inputId = id || generatedId
    const hasError = Boolean(error)
    const internalRef = React.useRef<HTMLDivElement>(null)
    const [internalValue, setInternalValue] = React.useState(defaultValue || '')
    const isControlled = value !== undefined
    const currentValue = isControlled ? value : internalValue

    // Combine refs
    React.useImperativeHandle(ref, () => internalRef.current!)

    // Sync contentEditable with value prop
    React.useEffect(() => {
      if (
        internalRef.current &&
        internalRef.current.textContent !== currentValue
      ) {
        internalRef.current.textContent = currentValue
      }
    }, [currentValue])

    // Get the appropriate state variant
    const state = getTextAreaState(disabled, readOnly, hasError)

    // Handle placeholder visibility
    const showPlaceholder = !currentValue && placeholder

    const handleInput = (e: React.FormEvent<HTMLDivElement>) => {
      const newValue = e.currentTarget.textContent || ''

      if (!isControlled) {
        setInternalValue(newValue)
      }

      onChange?.(newValue)
    }

    const handleKeyDown = (e: React.KeyboardEvent<HTMLDivElement>) => {
      // Prevent input when disabled or readOnly
      if (disabled || readOnly) {
        e.preventDefault()
        return
      }

      onKeyDown?.(e)
    }

    const handleLabelClick = () => {
      internalRef.current?.focus()
    }

    return (
      <div>
        {label && (
          <label
            htmlFor={inputId}
            className='block text-sm font-medium text-foreground mb-2 cursor-pointer'
            onClick={handleLabelClick}
          >
            {label}
            {required && (
              <span className='text-danger ml-1' aria-label='required'>
                *
              </span>
            )}
          </label>
        )}
        <div className={cn('relative w-full', className)}>
          <div
            ref={internalRef}
            id={inputId}
            role='textbox'
            aria-multiline='true'
            aria-invalid={hasError}
            aria-disabled={disabled}
            aria-readonly={readOnly}
            aria-required={required}
            aria-label={typeof label === 'string' ? label : undefined}
            contentEditable={!disabled && !readOnly}
            data-name={name}
            onInput={handleInput}
            onKeyDown={handleKeyDown}
            onPaste={e => handlePlainTextPaste(e, disabled, readOnly)}
            onBlur={onBlur}
            onFocus={onFocus}
            suppressContentEditableWarning
            style={{ minHeight: `${row * 1.5 + 1}rem` }}
            className={cn(
              textAreaInputVariants({ state }),
              'px-3 py-2',
              showPlaceholder &&
                'empty:before:content-[attr(data-placeholder)] empty:before:text-gray-400'
            )}
            data-placeholder={placeholder}
            {...props}
          />
        </div>
        {hasError && (
          <div className='mt-1 text-sm text-danger' role='alert'>
            {error}
          </div>
        )}
        {!hasError && helperText && (
          <div className='mt-1 text-sm text-gray-600'>{helperText}</div>
        )}
      </div>
    )
  }
)

TextAreaInput.displayName = 'TextAreaInput'
