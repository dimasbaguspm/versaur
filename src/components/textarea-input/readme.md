# TextAreaInput

A versatile textarea component that supports auto-resizing, configurable rows, and all Versaur color variants for multi-line text input.

## Features

- **Auto-resizing**: Use `fieldSizing="content"` for automatic height adjustment based on content
- **Configurable rows**: Set minimum and maximum rows for consistent sizing
- **Smart resize behavior**: Manual resize handle for fixed mode, no resize for auto mode
- **Full variant support**: All Versaur color variants (primary, secondary, tertiary, ghost, neutral) and semantic variants (success, info, warning, danger)
- **Outline variants**: Subtle styling options for each color
- **Error states**: Built-in error message display with proper ARIA attributes
- **Helper text**: Optional guidance text below the input
- **Accessibility**: Full WCAG 2.1 AA compliance with proper ARIA attributes

## Usage

```tsx
import { TextAreaInput } from '@/components/textarea-input'

// Basic usage
<TextAreaInput
  label="Description"
  placeholder="Enter your description..."
/>

// Auto-resizing textarea
<TextAreaInput
  label="Comments"
  fieldSizing="content"
  placeholder="Type your comments..."
/>

// Custom row configuration
<TextAreaInput
  label="Notes"
  minRows={4}
  maxRows={8}
  placeholder="Enter detailed notes..."
/>

// With error state
<TextAreaInput
  label="Message"
  error="This field is required"
  placeholder="Enter your message..."
/>

// With helper text
<TextAreaInput
  label="Feedback"
  helperText="Please provide detailed feedback to help us improve"
  placeholder="Share your thoughts..."
/>
```

## Props

| Prop | Type | Default | Description |
|------|------|---------|-------------|
| `label` | `ReactNode` | - | **Required.** Label text displayed above the textarea |
| `variant` | `TextAreaVariant` | `'primary'` | Visual style variant from Versaur design system |
| `fieldSizing` | `'content' \| 'fixed'` | `'fixed'` | Enable auto-resizing with CSS field-sizing property |
| `minRows` | `number` | `3` | Minimum number of rows for the textarea |
| `maxRows` | `number` | `5` | Maximum number of rows (only applies when fieldSizing is 'fixed') |
| `helperText` | `ReactNode` | - | Helper text displayed below the textarea |
| `error` | `ReactNode` | - | Error message for validation states |
| `disabled` | `boolean` | `false` | Disable the textarea |
| `className` | `string` | - | Additional CSS classes for the container |

Extends all standard HTML `textarea` attributes.

## Variants

### Core Variants
- `primary` - Coral accent (main actions)
- `secondary` - Sage accent (secondary actions)
- `tertiary` - Mist accent (subtle backgrounds)
- `ghost` - Slate accent (minimal styling)
- `neutral` - Light gray (neutral contexts)

### Semantic Variants
- `success` - Green accent (positive feedback)
- `info` - Blue accent (informational)
- `warning` - Orange accent (caution)
- `danger` - Red accent (errors)

### Outline Variants
Add `-outline` suffix to any variant for border-only styling (e.g., `primary-outline`).

## Field Sizing & Resize Behavior

The `fieldSizing` prop controls how the textarea behaves:

- **`fixed` (default)**: 
  - Static height based on `minRows`
  - Can be manually resized with resize handle (resize-y)
  - Respects `maxRows` for maximum height constraint
  
- **`content`**: 
  - Automatically grows/shrinks based on content
  - No manual resize handle (resize-none)
  - Ignores `maxRows` setting for unlimited growth

## Row Configuration

- **`minRows`**: Starting number of visible rows (default: 3)
- **`maxRows`**: Maximum rows when manually resizing in fixed mode (default: 5)

When `fieldSizing="content"`, the textarea starts at `minRows` and grows automatically with content.

## Accessibility

- Proper label association with `htmlFor` and `id`
- ARIA attributes (`aria-invalid`, `aria-disabled`)
- Error messages with `role="alert"`
- Focus management and keyboard navigation
- Meets WCAG 2.1 AA contrast requirements

## Examples

### Contact Form
```tsx
<TextAreaInput
  label="Message"
  placeholder="Tell us about your project..."
  minRows={4}
  maxRows={8}
  helperText="Please provide as much detail as possible"
/>
```

### Auto-resizing Comment Box
```tsx
<TextAreaInput
  label="Comments"
  fieldSizing="content"
  minRows={2}
  placeholder="Share your thoughts..."
  helperText="This field will auto-resize as you type"
/>
```

### Error State
```tsx
<TextAreaInput
  label="Description"
  error="Description must be at least 10 characters long"
  placeholder="Enter description..."
  minRows={3}
  maxRows={6}
/>
```
