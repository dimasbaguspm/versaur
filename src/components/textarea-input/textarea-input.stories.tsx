import type { Meta, StoryObj } from '@storybook/react'
import { TextAreaInput } from './textarea-input'

/**
 * TextAreaInput provides a styled textarea field for multi-line text input.
 * It supports field-sizing for auto-resizing, configurable minimum height,
 * and all the variant styles from the Versaur design system.
 */
const meta: Meta<typeof TextAreaInput> = {
  title: 'Form/TextAreaInput',
  component: TextAreaInput,
  parameters: {
    docs: {
      description: {
        component: `
A versatile textarea component that supports:
- Auto-resizing with field-sizing CSS property
- Configurable minimum height
- All Versaur color variants (primary, secondary, tertiary, ghost, neutral)
- Semantic variants (success, info, warning, danger)
- Outline variants for each color
- Error states and helper text
- Full accessibility support

Perfect for forms, comments, descriptions, and any multi-line text input needs.
        `,
      },
    },
  },
  args: {
    label: 'Description',
    placeholder: 'Enter your description here...',
  },
  argTypes: {
    variant: {
      control: 'select',
      options: [
        'primary',
        'primary-outline',
        'secondary',
        'secondary-outline',
        'tertiary',
        'tertiary-outline',
        'ghost',
        'ghost-outline',
        'neutral',
        'neutral-outline',
        'success',
        'success-outline',
        'info',
        'info-outline',
        'warning',
        'warning-outline',
        'danger',
        'danger-outline',
      ],
    },
    fieldSizing: {
      control: 'select',
      options: ['fixed', 'content'],
    },
    minRows: {
      control: 'number',
    },
    maxRows: {
      control: 'number',
    },
    disabled: {
      control: 'boolean',
    },
  },
}

export default meta
type Story = StoryObj<typeof meta>

/**
 * Default textarea with primary variant and fixed sizing
 */
export const Default: Story = {}

/**
 * Auto-resizing textarea that grows with content using field-sizing: content
 */
export const AutoResize: Story = {
  args: {
    label: 'Auto-resizing Comment',
    fieldSizing: 'content',
    placeholder: 'Type your message and watch it auto-resize...',
    defaultValue:
      'This textarea will automatically resize as you type more content.\n\nTry adding more lines to see it grow!',
  },
}

/**
 * Custom row configuration for larger text areas
 */
export const CustomRows: Story = {
  args: {
    label: 'Large Text Area',
    minRows: 5,
    maxRows: 8,
    placeholder: 'This textarea starts with 5 rows and can expand to 8 rows',
  },
}

/**
 * All color variants showcasing the Versaur design system
 */
export const Variants: Story = {
  render: () => (
    <div className='space-y-6 w-96'>
      <TextAreaInput
        label='Primary (Coral)'
        variant='primary'
        placeholder='Primary variant...'
      />
      <TextAreaInput
        label='Secondary (Sage)'
        variant='secondary'
        placeholder='Secondary variant...'
      />
      <TextAreaInput
        label='Tertiary (Mist)'
        variant='tertiary'
        placeholder='Tertiary variant...'
      />
      <TextAreaInput
        label='Ghost (Slate)'
        variant='ghost'
        placeholder='Ghost variant...'
      />
      <TextAreaInput
        label='Neutral'
        variant='neutral'
        placeholder='Neutral variant...'
      />
    </div>
  ),
}

/**
 * Outline variants for subtle styling
 */
export const OutlineVariants: Story = {
  render: () => (
    <div className='space-y-6 w-96'>
      <TextAreaInput
        label='Primary Outline'
        variant='primary-outline'
        placeholder='Primary outline variant...'
      />
      <TextAreaInput
        label='Secondary Outline'
        variant='secondary-outline'
        placeholder='Secondary outline variant...'
      />
      <TextAreaInput
        label='Ghost Outline'
        variant='ghost-outline'
        placeholder='Ghost outline variant...'
      />
    </div>
  ),
}

/**
 * Semantic variants for different states and contexts
 */
export const SemanticVariants: Story = {
  render: () => (
    <div className='space-y-6 w-96'>
      <TextAreaInput
        label='Success'
        variant='success'
        placeholder='Success state...'
        helperText='Great! Your input looks good.'
      />
      <TextAreaInput
        label='Info'
        variant='info'
        placeholder='Info state...'
        helperText='Additional information about this field.'
      />
      <TextAreaInput
        label='Warning'
        variant='warning'
        placeholder='Warning state...'
        helperText='Please review your input carefully.'
      />
      <TextAreaInput
        label='Danger'
        variant='danger'
        placeholder='Error state...'
        error='This field is required and cannot be empty.'
      />
    </div>
  ),
}

/**
 * Error state with validation message
 */
export const WithError: Story = {
  args: {
    label: 'Project Description',
    placeholder: 'Describe your project...',
    error: 'Description must be at least 10 characters long.',
    defaultValue: 'Too short',
  },
}

/**
 * With helper text for additional guidance
 */
export const WithHelperText: Story = {
  args: {
    label: 'Feedback',
    placeholder: 'Share your thoughts...',
    helperText:
      'Your feedback helps us improve our services. Please be as detailed as possible.',
  },
}

/**
 * Disabled state for read-only scenarios
 */
export const Disabled: Story = {
  args: {
    label: 'Read-only Content',
    disabled: true,
    defaultValue: 'This content cannot be edited in the current context.',
  },
}

/**
 * Form example showing practical usage in a contact form
 */
export const ContactForm: Story = {
  render: () => (
    <div className='space-y-6 w-96'>
      <TextAreaInput
        label='Message'
        placeholder='Tell us about your project...'
        minRows={4}
        maxRows={8}
        helperText='Please provide as much detail as possible about your requirements.'
      />
      <TextAreaInput
        label='Additional Notes'
        variant='secondary'
        fieldSizing='content'
        placeholder='Any additional information...'
        helperText='This field will auto-resize as you type.'
      />
    </div>
  ),
}
