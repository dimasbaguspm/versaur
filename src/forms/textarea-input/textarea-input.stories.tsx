import type { Meta, StoryObj } from '@storybook/react'
import { TextAreaInput } from './textarea-input'

/**
 * TextAreaInput provides a robust multi-line text input using contentEditable div.
 * Uses primary variant by default with support for error states, disabled, and readOnly modes.
 */
const meta: Meta<typeof TextAreaInput> = {
  title: 'Forms/TextAreaInput',
  component: TextAreaInput,
  parameters: {
    docs: {
      description: {
        component: `
A versatile textarea component built with contentEditable div for better control:
- Uses primary variant styling by default
- Error states with validation messages
- Disabled and readOnly support
- Helper text for guidance
- Full accessibility with proper ARIA attributes
- Works cross-browser with consistent behavior

Perfect for forms, comments, descriptions, and any multi-line text input needs.
        `,
      },
    },
  },
  args: {
    label: 'Description',
    placeholder: 'Enter your description here...',
  },
  argTypes: {
    disabled: {
      control: 'boolean',
    },
    readOnly: {
      control: 'boolean',
    },
    row: {
      control: 'number',
      description: 'Height in rem units',
    },
  },
}

export default meta
type Story = StoryObj<typeof meta>

/**
 * Default textarea with primary styling
 */
export const Default: Story = {}

/**
 * With default value for uncontrolled usage
 */
export const WithDefaultValue: Story = {
  args: {
    label: 'Comment',
    placeholder: 'Add your comment...',
    defaultValue:
      'This is a default value that can be edited.\n\nIt supports multi-line content naturally.',
  },
}

/**
 * Error state with validation message
 */
export const WithError: Story = {
  args: {
    label: 'Project Description',
    placeholder: 'Describe your project...',
    error: 'Description must be at least 10 characters long.',
    defaultValue: 'Too short',
  },
}

/**
 * With helper text for additional guidance
 */
export const WithHelperText: Story = {
  args: {
    label: 'Feedback',
    placeholder: 'Share your thoughts...',
    helperText:
      'Your feedback helps us improve our services. Please be as detailed as possible.',
  },
}

/**
 * Required field with asterisk indicator
 */
export const Required: Story = {
  args: {
    label: 'Message',
    placeholder: 'Enter your message...',
    required: true,
    helperText: 'This field is required and must be filled out.',
  },
}

/**
 * Disabled state prevents all interactions
 */
export const Disabled: Story = {
  args: {
    label: 'Disabled Field',
    disabled: true,
    defaultValue:
      'This content cannot be edited because the field is disabled.',
  },
}

/**
 * ReadOnly state allows viewing but not editing
 */
export const ReadOnly: Story = {
  args: {
    label: 'Read-Only Content',
    readOnly: true,
    defaultValue:
      'This content is read-only. You can select and copy the text, but cannot modify it.',
  },
}

/**
 * Custom height using row prop (in rem units)
 */
export const CustomHeight: Story = {
  render: () => (
    <div className='space-y-6 w-96'>
      <TextAreaInput
        label='Small (1 row)'
        placeholder='Compact textarea...'
        row={1}
      />
      <TextAreaInput
        label='Default (3 rows)'
        placeholder='Default size textarea...'
      />
      <TextAreaInput
        label='Large (10 rows)'
        placeholder='Expanded textarea...'
        row={10}
      />
    </div>
  ),
}
