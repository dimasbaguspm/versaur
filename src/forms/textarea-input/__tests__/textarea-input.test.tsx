import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { composeStories } from '@storybook/react'
import { describe, it, expect } from 'vitest'
import * as stories from '../textarea-input.stories'

const { Default, WithError, Disabled, ReadOnly, WithDefaultValue } =
  composeStories(stories)

describe('TextAreaInput', () => {
  describe('Rendering', () => {
    it('renders with label and placeholder', () => {
      render(<Default />)
      expect(screen.getByLabelText('Description')).toBeInTheDocument()
      const textbox = screen.getByRole('textbox', { name: 'Description' })
      expect(textbox).toBeInTheDocument()
      expect(textbox).toHaveAttribute(
        'data-placeholder',
        'Enter your description here...'
      )
    })

    it('matches snapshot', () => {
      const { asFragment } = render(<Default />)
      expect(asFragment()).toMatchSnapshot()
    })
  })

  describe('Basic Functionality', () => {
    it('accepts user input via contentEditable', async () => {
      const user = userEvent.setup()
      render(<Default />)
      const textbox = screen.getByRole('textbox')

      await user.click(textbox)
      await user.keyboard('This is test content')

      expect(textbox.textContent).toBe('This is test content')
    })

    it('displays default value', () => {
      render(<WithDefaultValue />)
      const textbox = screen.getByRole('textbox')
      expect(textbox.textContent).toContain('This is a default value')
    })
  })

  describe('Error States', () => {
    it('displays error message when error prop is provided', () => {
      render(<WithError />)
      const errorMessage = screen.getByRole('alert')
      expect(errorMessage).toBeInTheDocument()
      expect(errorMessage).toHaveTextContent(
        'Description must be at least 10 characters long.'
      )
    })

    it('applies error styling when error is present', () => {
      render(<WithError />)
      const textbox = screen.getByRole('textbox')
      expect(textbox).toHaveAttribute('aria-invalid', 'true')
    })
  })

  describe('Disabled State', () => {
    it('sets aria-disabled when disabled prop is true', () => {
      render(<Disabled />)
      const textbox = screen.getByRole('textbox')
      expect(textbox).toHaveAttribute('aria-disabled', 'true')
      expect(textbox).toHaveAttribute('contenteditable', 'false')
    })

    it('prevents input when disabled', async () => {
      const user = userEvent.setup()
      render(<Disabled />)
      const textbox = screen.getByRole('textbox')
      const originalContent = textbox.textContent

      await user.click(textbox)
      await user.keyboard('Try to type')

      expect(textbox.textContent).toBe(originalContent)
    })
  })

  describe('ReadOnly State', () => {
    it('sets aria-readonly when readOnly prop is true', () => {
      render(<ReadOnly />)
      const textbox = screen.getByRole('textbox')
      expect(textbox).toHaveAttribute('aria-readonly', 'true')
      expect(textbox).toHaveAttribute('contenteditable', 'false')
    })

    it('prevents input when readOnly', async () => {
      const user = userEvent.setup()
      render(<ReadOnly />)
      const textbox = screen.getByRole('textbox')
      const originalContent = textbox.textContent

      await user.click(textbox)
      await user.keyboard('Try to type')

      expect(textbox.textContent).toBe(originalContent)
    })
  })

  describe('Accessibility', () => {
    it('has proper ARIA attributes', () => {
      render(<Default />)
      const textbox = screen.getByRole('textbox')
      expect(textbox).toHaveAttribute('aria-multiline', 'true')
      expect(textbox).toHaveAttribute('aria-invalid', 'false')
    })

    it('displays helper text when provided', () => {
      const { container } = render(<Default helperText='This is helper text' />)
      expect(container).toHaveTextContent('This is helper text')
    })
  })

  describe('Custom Row Height', () => {
    it('applies custom row height calculation', () => {
      const { container } = render(<Default row={10} />)
      const textbox = container.querySelector('[role="textbox"]')
      // Formula: row * 1.5 + 1 = 10 * 1.5 + 1 = 16rem
      expect(textbox).toHaveStyle({ minHeight: '16rem' })
    })

    it('uses default height calculation when row prop is not provided', () => {
      const { container } = render(<Default />)
      const textbox = container.querySelector('[role="textbox"]')
      // Formula: 3 * 1.5 + 1 = 5.5rem (default row is 3)
      expect(textbox).toHaveStyle({ minHeight: '5.5rem' })
    })
  })

  describe('Required State', () => {
    it('displays asterisk when required prop is true', () => {
      const { container } = render(<Default label='Message' required />)
      const asterisk = container.querySelector('[aria-label="required"]')
      expect(asterisk).toBeInTheDocument()
      expect(asterisk).toHaveTextContent('*')
      expect(asterisk).toHaveClass('text-danger')
    })

    it('sets aria-required attribute on textbox', () => {
      render(<Default required />)
      const textbox = screen.getByRole('textbox')
      expect(textbox).toHaveAttribute('aria-required', 'true')
    })

    it('does not display asterisk when required is false', () => {
      const { container } = render(<Default label='Message' required={false} />)
      const asterisk = container.querySelector('[aria-label="required"]')
      expect(asterisk).not.toBeInTheDocument()
    })
  })
})
