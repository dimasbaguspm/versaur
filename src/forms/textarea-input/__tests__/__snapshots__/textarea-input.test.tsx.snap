// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TextAreaInput > Rendering > matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <label
      class="block text-sm font-medium text-foreground mb-2 cursor-pointer"
      for="«r1»"
    >
      Description
    </label>
    <div
      class="relative w-full"
    >
      <div
        aria-invalid="false"
        aria-label="Description"
        aria-multiline="true"
        class="block w-full rounded-md border bg-white transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:pointer-events-none disabled:bg-gray-50 overflow-y-auto whitespace-pre-wrap break-words border-primary/30 text-foreground focus:border-primary focus:ring-primary/20 px-3 py-2 empty:before:content-[attr(data-placeholder)] empty:before:text-gray-400"
        contenteditable="true"
        data-placeholder="Enter your description here..."
        id="«r1»"
        role="textbox"
        style="min-height: 5.5rem;"
      />
    </div>
  </div>
</DocumentFragment>
`;
